### PR TITLE
[core] node affinity soft chooses another node if it is not available

### PR DESCRIPTION
**Pull request by clarng**
## Why are these changes needed?

Addressed https://github.com/ray-project/ray/issues/33835 to enable spreading of dataset tasks

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/raylet/scheduling/policy/node_affinity_scheduling_policy.cc [hunk 1]
```diff
@@ -24,7 +24,13 @@ scheduling::NodeID NodeAffinitySchedulingPolicy::Schedule(
   scheduling::NodeID target_node_id = scheduling::NodeID(options.node_affinity_node_id);
   if (nodes_.contains(target_node_id) && is_node_alive_(target_node_id) &&
       nodes_.at(target_node_id).GetLocalView().IsFeasible(resource_request)) {
```
Comment (ericl) [inline src/ray/raylet/scheduling/policy/node_affinity_scheduling_policy.cc:26]: The logic looks good, but I feel we could clean up the if statements a little. The nested if / return is a bit hard to parse.

Comment (clarng) [reply src/ray/raylet/scheduling/policy/node_affinity_scheduling_policy.cc]: ditto :)

**Review by ericl — APPROVED** [review]

**Review by jjyao — CHANGES_REQUESTED** [review]
Could you also
1. Run push based shuffle release tests to make sure there is no regression
2. Update the doc to reflect the new behavior

**Comment by clarng** [comment]
Release test so far shows no regression

https://buildkite.com/ray-project/release-tests-pr/builds/33077

PR latency vs baseline latency
push_based_random_shuffle_1tb       876 vs  869 last run 
push_based_sort_1tb     2070 vs. 2443 last run (6737 max)

chaos_push_based_random_shuffle_1tb.     1738 vs 2752 last run (5275 max)

**Comment by jjyao** [comment]
@clarng could you also double check `soft=True` usages in the codebase and make sure the new behavior works as well. I saw some in data and some in serve.

**Comment by clarng** [comment]
> @clarng could you also double check `soft=True` usages in the codebase and make sure the new behavior works as well. I saw some in data and some in serve.

Quick audit shows it looks fine to me : https://github.com/ray-project/ray/issues/33835#issuecomment-1489362894

In fact, I think they would prefer liveness over strict scheduling for where it is used today, and should benefit from this change

**Review by jjyao — APPROVED** [review]

**Comment by ericl** [comment]
@zhe-thoughts this is a blocker for streaming execution

**Review by zhe-thoughts — APPROVED** [review]
OK let's merge since it's a blocker
